pandas/core/describe.py

```diff
@@ -60,107 +61,155 @@ def describe_ndframe(
+    describer = create_describer(
+        obj,
+        include=include,
+        exclude=exclude,
+        datetime_is_numeric=datetime_is_numeric,
+    )
+    result = describer.describe(percentiles=percentiles)
+    return cast(FrameOrSeries, result)
+
 
+def create_describer(
+    obj: "FrameOrSeries",
+    include: Optional[Union[str, Sequence[str]]],
+    exclude: Optional[Union[str, Sequence[str]]],
+    datetime_is_numeric: bool,
+) -> "NDFrameDescriberAbstract":
+    """
+    Create concrete NDFrameDescriberAbstract instance suitable for the object.
+    """
+    describer: NDFrameDescriberAbstract
     if obj.ndim == 1:
-        result_series = describe_series(
-            cast("Series", obj),
-            percentiles,
-            datetime_is_numeric,
+        return SeriesDescriber(
+            series=cast("Series", obj),
+            datetime_is_numeric=datetime_is_numeric,
+        )
+    else:
+        return DataFrameDescriber(
+            frame=cast("DataFrame", obj),
+            include=include,
+            exclude=exclude,
+            datetime_is_numeric=datetime_is_numeric,
         )
-        return cast(FrameOrSeries, result_series)
 
-    frame = cast("DataFrame", obj)
 
-    if frame.ndim == 2 and frame.columns.size == 0:
-        raise ValueError("Cannot describe a DataFrame without columns")
+class NDFrameDescriberAbstract(ABC):
+    """Abstract class for describing dataframe or series."""
 
-    result_frame = describe_frame(
-        frame=frame,
-        include=include,
-        exclude=exclude,
-        percentiles=percentiles,
-        datetime_is_numeric=datetime_is_numeric,
-    )
-    return cast(FrameOrSeries, result_frame)
+    @abstractmethod
+    def describe(self, percentiles: Sequence[float]) -> "FrameOrSeriesUnion":
+        """Do describe either series or dataframe.
 
+        Parameters
+        ----------
+        percentiles : list-like of numbers
+            The percentiles to include in the output.
+        """
 
-def describe_series(
-    series: "Series",
-    percentiles: Sequence[float],
-    datetime_is_numeric: bool,
-) -> "Series":
-    """Describe series.
 
-    The reason for the delegation to ``describe_1d`` only:
-    to allow for a proper stacklevel of the FutureWarning.
+class SeriesDescriber(NDFrameDescriberAbstract):
+    """Class responsible for creating series description.
 
     Parameters
     ----------
-    series : Series
+    data : Series
         Series to be described.
-    percentiles : list-like of numbers
-        The percentiles to include in the output.
-    datetime_is_numeric : bool, default False
+    datetime_is_numeric : bool
         Whether to treat datetime dtypes as numeric.
-
-    Returns
-    -------
-    Series
     """
-    return describe_1d(
-        series,
-        percentiles,
-        datetime_is_numeric,
-        is_series=True,
-    )
 
+    def __init__(
+        self,
+        series: "Series",
```

Comment:
can you make this a common method in NDFrame (and changes series -> data) to match dataframe

Reply:
I will check that. But I guess it is going to violate Liskov principle.

Reply:
Looks like nothing is violated. It seems to work.